arch/sim/src/sim/sim_netdriver.c

```diff
@@ -66,288 +66,161 @@
-                {
-                  ninfo("IPv4 frame\n");
-                  NETDEV_RXIPV4(dev);
-
-                  /* Receive an IPv4 packet from the network device */
-
-                  ipv4_input(dev);
-
-                  /* Check for a reply to the IPv4 packet */
-
-                  netdriver_reply(dev);
-                }
-              else
-#endif /* CONFIG_NET_IPv4 */
-#ifdef CONFIG_NET_IPv6
-              if (eth->type == HTONS(ETHTYPE_IP6))
-                {
-                  ninfo("IPv6 frame\n");
-                  NETDEV_RXIPV6(dev);
-
-                  /* Give the IPv6 packet to the network layer */
-
-                  ipv6_input(dev);
-
-                  /* Check for a reply to the IPv6 packet */
-
-                  netdriver_reply(dev);
-                }
-              else
-#endif/* CONFIG_NET_IPv6 */
-#ifdef CONFIG_NET_ARP
-              if (eth->type == HTONS(ETHTYPE_ARP))
-                {
-                  ninfo("ARP frame\n");
-                  NETDEV_RXARP(dev);
-
-                  arp_input(dev);
-
-                  /* If the above function invocation resulted in data that
-                   * should be sent out on the network, the global variable
-                   * d_len is set to a value > 0.
-                   */
-
-                  if (dev->d_len > 0)
-                    {
-                      netdriver_send(dev);
-                    }
-                }
-              else
+#ifdef SIM_NETDEV_RECV_OFFLOAD
+      len = sim_netdev_read(DEVIDX(priv), netpkt_getbase(pkt),
+                            SIM_NETDEV_BUFSIZE);
+#else
+      len = sim_netdev_read(DEVIDX(priv), priv->buf, SIM_NETDEV_BUFSIZE);
 #endif
-                {
-                  NETDEV_RXDROPPED(dev);
-                  nwarn("WARNING: Unsupported Ethernet type %u\n",
-                        eth->type);
-                }
-            }
-          else
-            {
-              NETDEV_RXERRORS(dev);
-            }
+      if (len == 0)
+        {
+          netpkt_free(dev, pkt, NETPKT_RX);
+          return NULL;
         }
 
-#ifdef SIM_NETDEV_IOB_OFFLOAD
-      netdev_iob_release(dev);
+#ifdef SIM_NETDEV_RECV_OFFLOAD
+      netpkt_reset_reserved(dev, pkt, 0); /* No overhead before data. */
+      netpkt_setdatalen(dev, pkt, len);
+#else
+      netpkt_copyin(dev, pkt, priv->buf, len, 0);
 #endif
     }
 
-  net_unlock();
+  return pkt;
 }
 
-static int netdriver_txpoll(struct net_driver_s *dev)
+static int netdriver_ifup(struct netdev_lowerhalf_s *dev)
 {
-  /* Send the packet */
-
-  NETDEV_TXPACKETS(dev);
-  netdriver_send(dev);
-  NETDEV_TXDONE(dev);
+  struct sim_netdev_s *priv = (struct sim_netdev_s *)dev;
 
-  /* If zero is returned, the polling will continue until all connections
-   * have been examined.
-   */
-
-  return 0;
-}
-
-static int netdriver_ifup(struct net_driver_s *dev)
-{
-  int devidx = (intptr_t)dev->d_private;
-
-  UNUSED(devidx);
+  UNUSED(priv);
```

Comment:
Why do we need `priv`? We can use `dev` directly since `DEVIDX` is already doing a necessary type cast

Reply:
Done. Also removed `priv` from other functions (add a macro `DEVBUF`)